Fix restore when a package is installed with a version specified in CPM

vernou opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->

# Bug

<!-- If this is an engineering change or test change only, you do not need an issue. -->
<!-- Find or create an issue in NuGet/Home and paste the full url. -->
<!-- At the maintainers discretion, multiple changes may apply to a single issue, but only when the PRs are all created within a short period of time. -->
Fixes: https://github.com/NuGet/Home/issues/13657

## Description

When `dotnet add package` with version specified in `Directory.Packages.props`, the restored dependency is the last version and no the version specified in `Directory.Packages.props`.

The modification :
If a CPM version is available for the installed package then uses it.
Else old behavior (last package version)

## PR Checklist

- [x] Meaningful title, helpful description and a linked NuGet/Home issue
- [ ] Added tests
- [ ] Link to an issue or pull request to update docs if this PR changes settings, environment variables, new feature, etc.

vernou committed:
Fix

zivkan commented:
@vernou thanks for the PR. I was hoping it'd be as easy as what this PR currently is.

If you look at the CI build, you'll see there's 2 failing test across multiple platforms: AddPkg_WithCPM_WithPackageReference_WithVersionOverride_WithPackageVersion_NoVersionCLI_NoOps and AddPkg_WithCPM_WhenPackageVersionDoesNotExistAndVersionCLIArgNotPassed_Success

Have a look at what other tests also exist in the same class, and figure out if we need any more test scenarios.

The other tests failing on CI look like they're flakey, so ignore them.

vernou commented:
The test `AddPkg_WithCPM_WithPackageReference_WithVersionOverride_WithPackageVersion_NoVersionCLI_NoOps` expects the command `add package` fails.
The commande fails because the resolved version range from source hasn't `OriginalString`, so MSBuild fail to update the project with this error :
```
error: System.AggregateException: One or more errors occurred. (Parameter "Value" cannot be null.)
error:  ---> System.ArgumentNullException: Parameter "Value" cannot be null.
error:    at Microsoft.Build.Shared.ErrorUtilities.ThrowArgumentNull(String parameterName, String resourceName)
error:    at Microsoft.Build.Construction.ProjectMetadataElement.set_Value(String value)
error:    at NuGet.CommandLine.XPlat.MSBuildAPIUtility.UpdateVersionOverride(Project project, ProjectItem packageReference, String versionCLIArgument) in \NuGet.Client\src\NuGet.Core\NuGet.CommandLine.XPlat\Utility\MSBuildAPIUtility.cs:line 575
error:    at NuGet.CommandLine.XPlat.MSBuildAPIUtility.AddPackageReference(Project project, LibraryDependency libraryDependency, IEnumerable`1 existingPackageReferences, Boolean noVersion, String framework) in \NuGet.Client\src\NuGet.Core\NuGet.CommandLine.XPlat\Utility\MSBuildAPIUtility.cs:line 332
error:    at NuGet.CommandLine.XPlat.MSBuildAPIUtility.AddPackageReference(String projectPath, LibraryDependency libraryDependency, Boolean noVersion) in \NuGet.Client\src\NuGet.Core\NuGet.CommandLine.XPlat\Utility\MSBuildAPIUtility.cs:line 258
error:    at NuGet.CommandLine.XPlat.AddPackageReferenceCommandRunner.ExecuteCommand(PackageReferenceArgs packageReferenceArgs, MSBuildAPIUtility msBuild) in \NuGet.Client\src\NuGet.Core\NuGet.CommandLine.XPlat\Commands\PackageReferenceCommands\AddPackageReferenceCommandRunner.cs:line 242
```

With the PR fix, the resolved version range from CPM has `OriginalString`, so MSBuild success to update the project (but do nothing, because CPM is enable) and the command success... and the test fails.

---

I see two possibilities :
- Adapt the test
- Force to set `OriginalString` to null in the version range retrieved from the CPM

I don't undertand why the command is expected to fail and the reason/error sounds dirty.
So I think the test should be modified.

@zivkan, what do you think?

vernou committed:
Set `VersionRange.OriginalString`

vernou committed:
Fix when package version isn't specified in CPM

vernou commented:
I implemented the second to see. When the version is resolved from CPM, the resolved version `OriginalString` is forced to null.
So `AddPkg_WithCPM_WithPackageReference_WithVersionOverride_WithPackageVersion_NoVersionCLI_NoOps` success.

---

I fixed my code and `AddPkg_WithCPM_WhenPackageVersionDoesNotExistAndVersionCLIArgNotPassed_Success` success.

vernou committed:
Add test to reproduce the bug

vernou committed:
Keep central version range original string

vernou committed:
Restore removed trailling spaces to reduce change count

vernou commented:
After going through the tests to be able to add a test, I think I have a better understanding of how they work.
Finally I decided to keep central version range original string and alter the test `AddPkg_WithCPM_WithPackageReference_WithVersionOverride_WithPackageVersion_NoVersionCLI_NoOps`.

@zivkan, the PR is ready to review.

kartheekp-ms commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs:
```diff
@@ -132,22 +132,38 @@ public async Task<int> ExecuteCommand(PackageReferenceArgs packageReferenceArgs,
-                if (latestVersion == null)
+                if (originalPackageSpec.RestoreMetadata.CentralPackageVersionsEnabled)
+                {
+                    Thread.Sleep(TimeSpan.FromSeconds(10));
```
I am curious, why do we have to pause the execution for 10 seconds?

vernou replied:
A trick to debug unit test. The pause help me to attach the process stated by the test.
It's a mistake, nice catch.

vernou replied:
It's removed.

vernou committed:
Remove bad `Thread.Sleep`

kartheekp-ms commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs:
```diff
@@ -132,22 +132,37 @@ public async Task<int> ExecuteCommand(PackageReferenceArgs packageReferenceArgs,
+                        .TargetFrameworks
+                        .Where(tf => tf.CentralPackageVersions.ContainsKey(packageReferenceArgs.PackageId))
+                        .Select(tf => tf.CentralPackageVersions[packageReferenceArgs.PackageId])
+                        .Max();
```
`Max()` is likely unnecessary since CPM settings should specify a single version for a package. If multiple versions are defined, MSBuild typically selects the last one, ensuring a single version remains after evaluation. Please let me know if I am missing something.

vernou replied:
I don't know how MSBuild generate the dg. I see a collection, so I choose the max version because when a dependency has many reference it's the max that is restored.

I followed your guidance and I replaced `Max` by `FirstOrDefault`

nkolev92 replied:
CentralPackageVersions is a dictionary, so it'll always have 1 result

vernou committed:
Replace `Max` by `FirstOrDefault`
